deepinv/models/complex.py

```diff
@@ -0,0 +1,35 @@
+import torch
+
+
+def to_complex_denoiser(denoiser, mode="real_imag"):
+    r"""
+    Given a denoiser, returns a corresponding modified denoiser which can process complex numbers.
+
+    :param torch.nn.Module denoiser: a denoiser which takes in real-valued inputs.
+    :param str mode: the mode by which the complex inputs are processed. Can be either `'real_imag'` or `'abs_angle'`.
+    :return: (torch.nn.Module) the denoiser which takes in complex-valued inputs.
+    """
+
+    class complex_denoiser(torch.nn.Module):
+        def __init__(self, denoiser: torch.nn.Module, mode, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.mode = mode
+            self.denoiser = denoiser
+
+        def forward(self, x, sigma=None):
+            if self.mode == "real_imag":
+                x_real = x.real
+                x_imag = x.imag
+                x_real = self.denoiser.forward(x_real, sigma)
+                x_imag = self.denoiser.forward(x_imag, sigma)
```

Comment:
Is it possible to do a stacking over the batch dimension instead of doing 2 forward passes?

Reply:
Do you mean instead of processing x_real and x_imag separately, we first stack them over the first dimension and pass the new tensor into the denoiser? It is indeed a possible approach, but I think the space and computation requirements would be the same, @tachella do you think it is better?

Reply:
I meant something like this because I've witnessed it was faster in some cases, but I didn't do an extensive check

```python
        noisy_batch = torch.cat((x.real, x.imag), 0)
        denoised_batch = self.denoiser(noisy_batch, sigma)
        denoised = denoised_batch[0:x.real.shape[0], ...]+1j*denoised_batch[x.real.shape[0]:-1, ...]
```

Reply:
Thanks for the clarification! I did a small benchmark, it shows that the time of stacking and doing one forward pass is nearly the same (155.91s) as doing two forward passes (157.50s) for 100 batches of 5 random images, so I think the current implementation is fine.

Reply:
I've tested the following code on an NVIDIA GeForce RTX 3090 Ti

```
device = 'cuda:0'
x = torch.randn(16, 3, 128, 128, device=device) + torch.randn(16, 3, 128, 128, device=device) * 1j

drunet = dinv.models.DRUNet().to(device)

denoiser = to_complex_denoiser(drunet, mode="real_imag") # without cat (current solution)
denoiser2 = to_complex_denoiser2(drunet, mode="real_imag") # with cat

start = time.time()
for i in range(100):
    xhat = denoiser(x, 0.1)

print("Time 1: ", time.time() - start)

start = time.time()
for i in range(100):
    xhat = denoiser2(x, 0.1)

print("Time 2: ", time.time() - start)
```
and got 
Time 1:  9.287079334259033
Time 2:  7.886474609375

Reply:
an important detail, I had to modify very slightly @matthieutrs suggestion:

```
  noisy_batch = torch.cat((x.real, x.imag), 0)
  denoised_batch = self.denoiser(noisy_batch, sigma)
  return denoised_batch[:x.real.shape[0], ...] + 1j * denoised_batch[x.real.shape[0]:, ...]
```